### PR TITLE
Refactor to Typescript

### PR DESCRIPTION
**Pull request by xcambar**
Done with the help of:
* https://github.com/typed-ember/ember-cli-typescript
* https://ember-cli-typescript.com/docs

### CODE REVIEW

File: addon/initializers/embedded.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  | DeprecatedBooleanConfig
+  | ObjectConfig
+
+function configIsVoid(config:GivenConfig): config is VoidConfig {
+  return [null, undefined].includes(config as VoidConfig)
+}
+
+function configIsBoolean(config:GivenConfig): config is DeprecatedBooleanConfig {
+  return [true, false].includes(config as DeprecatedBooleanConfig)
+}
```
Comment (MrChocolatine) [inline addon/initializers/embedded.ts:27]: Can't we simply check `typeof config === 'boolean'` ?

File: addon/initializers/embedded.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import EmberApplication from '@ember/application'
+import EmberObject from '@ember/object'
+import { deprecate } from '@ember/application/deprecations';
```
[the remark this answers is not in the document]
Comment (GreatWizard) [reply addon/initializers/embedded.ts:3]:
```suggestion
import { deprecate } from '@ember/application/deprecations';
```

File: types/dummy/index.d.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (GreatWizard) [reply types/dummy/index.d.ts:1]: I guess this one is generated by ember-cli-ts

File: addon/initializers/embedded.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+type GivenConfig = VoidConfig
+  | DeprecatedBooleanConfig
+  | ObjectConfig
+
+function configIsVoid(config:GivenConfig): config is VoidConfig {
+  return [null, undefined].includes(config as VoidConfig)
+}
+
+function configIsBoolean(config:GivenConfig): config is DeprecatedBooleanConfig {
+  return [true, false].includes(config as DeprecatedBooleanConfig)
```
Comment (GreatWizard) [inline addon/initializers/embedded.ts:26]:
```suggestion
  return typeof config === 'boolean'
```

File: .travis.yml [hunk 5]
```diff
@@ -37,19 +36,28 @@ jobs:
...
     - env: EMBER_TRY_SCENARIO=ember-lts-3.12
+    - env: EMBER_TRY_SCENARIO=ember-lts-3.16
     - env: EMBER_TRY_SCENARIO=ember-release
     - env: EMBER_TRY_SCENARIO=ember-beta
     - env: EMBER_TRY_SCENARIO=ember-canary
     - env: EMBER_TRY_SCENARIO=ember-default-with-jquery
+    - env: EMBER_TRY_SCENARIO=ember-classic
+
+before_install:
+  - curl -o- -L https://yarnpkg.com/install.sh | bash
```
[the remark this answers is not in the document]
Comment (GreatWizard) [reply .travis.yml:56]: -o**-** means will output its response to stdout and not a file

**Comment by GreatWizard** [comment]
I think we should also convert dummy and tests in this PR

**Comment by xcambar** [comment]
@GreatWizard We may in the future convert tests to TS, but the goal og this PR was to test compatibility of an addon in TS with a JS-based app.
In that regard, the PR is ready, and I'd rather do a later PR for the test than overwork this one.

What do you think?

**Review by GreatWizard — APPROVED** [review]

**Pull request merged**